File: Drag&Drop/Demos/NormalCollectionViewController.m
//
//  NormalCollectionViewController.m
//  Drag&Drop
//
//  Created by Joey Patino on 11/4/15.
//  Copyright © 2015 Joseph Patino. All rights reserved.
//

#import "NormalCollectionViewController.h"
#import "UICollectionView+InteractiveMovementSupport.h"

@interface NormalCollectionViewController () <UICollectionViewDataSource, UICollectionViewDelegate, UICollectionViewDelegateFlowLayout>

@property (nonatomic, strong) UICollectionView *collection;
@property (nonatomic, strong) UICollectionViewFlowLayout *layout;
@property (nonatomic, strong) NSMutableArray *collectionSource;
@property (nonatomic, strong) NSMutableArray *collectionHeights;

@end

@implementation NormalCollectionViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    [self loadContent];
}

- (void)loadContent {
    
    self.collectionSource = [NSMutableArray array];
    self.collectionHeights = [NSMutableArray array];
    
    for (int i = 0; i < 300; i++) {
        [self.collectionSource addObject:@(i)];
        int n = arc4random() % 120;
        if (n < 40) n = 40;
        [self.collectionHeights addObject:@(n)];
    }
    

    self.layout = [[UICollectionViewFlowLayout alloc] init];
    self.layout.minimumInteritemSpacing = 4;
    self.layout.minimumLineSpacing = 4;
    self.layout.sectionInset = UIEdgeInsetsMake(4, 4, 4, 4);
    self.collection = [[UICollectionView alloc] initWithFrame:CGRectMake(0, 20,
                                                                         self.view.bounds.size.width,
                                                                         self.view.bounds.size.height-20)
                       
                                         collectionViewLayout:self.layout];
    self.collection.delegate = self;
    self.collection.dataSource = self;
    [self.view addSubview:self.collection];
    [self.collection registerClass:[UICollectionViewCell class] forCellWithReuseIdentifier:@"CollectionViewCell"];
    
    [self.collection reloadData];
}

#pragma mark -

- (NSInteger)collectionView:(UICollectionView *)collectionView numberOfItemsInSection:(NSInteger)section {
    return self.collectionSource.count;
}

- (NSInteger)numberOfSectionsInCollectionView:(UICollectionView *)collectionView {
    return 1;
}

- (UICollectionViewCell *)collectionView:(UICollectionView *)collectionView cellForItemAtIndexPath:(NSIndexPath *)indexPath {
    
    UICollectionViewCell *cell = [collectionView dequeueReusableCellWithReuseIdentifier:@"CollectionViewCell" forIndexPath:indexPath];
    cell.backgroundColor = [UIColor whiteColor];
    
    [cell.contentView.subviews makeObjectsPerformSelector:@selector(removeFromSuperview)];
    NSNumber *n = self.collectionSource[indexPath.row];
    [self applyLabel:[NSString stringWithFormat:@"%li", (long)n.integerValue] toView:cell.contentView];

    return cell;
}

- (void)collectionView:(UICollectionView *)collectionView willDisplayCell:(nonnull UICollectionViewCell *)cell forItemAtIndexPath:(nonnull NSIndexPath *)indexPath {
    [collectionView enableDragAndDropForCell:cell atIndexPath:indexPath];
}

- (CGSize)collectionView:(UICollectionView *)collectionView layout:(UICollectionViewLayout *)collectionViewLayout sizeForItemAtIndexPath:(NSIndexPath *)indexPath {
    
    return CGSizeMake((self.view.bounds.size.width/2)-6, [self.collectionHeights[indexPath.row] integerValue]);
}

- (void)collectionView:(UICollectionView *)collectionView moveItemAtIndexPath:(NSIndexPath *)sourceIndexPath toIndexPath:(NSIndexPath*)destinationIndexPath {
    
    NSObject *item = [self.collectionSource objectAtIndex:sourceIndexPath.row];
    [self.collectionSource removeObject:item];
    [self.collectionSource insertObject:item atIndex:destinationIndexPath.row];
}

- (void)applyLabel:(NSString *)string toView:(UIView *)view {
    UILabel *l = [[UILabel alloc] init];
    l.text = string;
    [l sizeToFit];
    l.frame = CGRectMake(0, 0, CGRectGetWidth(l.frame), CGRectGetHeight(l.frame));
    [view addSubview:l];
}

@end

File: Demo/DragDropDemo/Demos/NormalCollectionViewController.swift
//
//  NormalCollectionViewController.swift
//  DragDropDemo
//
//  Created by Joey Patino on 11/4/15.
//  Copyright © 2015 Joseph Patino. All rights reserved.
//

import UIKit
import DragDrop
import DemoKit

/// A curation grid: masonry cards reordered by dragging. One collection view,
/// one section, 300 items -- the reorder path at a scale where a wrong frame is
/// visible immediately.
final class NormalCollectionViewController: DemoViewController {

    private var collection: UICollectionView?
    private var cards: [Int] = []
    private var heights: [CGFloat] = []

    override func loadContent() {
        title = "Moodboard"

        cards = Array(0..<300)
        // Seeded, not random: the same position always gets the same height, so
        // a screenshot taken today matches one taken tomorrow, and a reorder
        // leaves the masonry rhythm alone rather than dragging a card's height
        // around with it.
        //
        // `stableMix` rather than `stableHash("card-\(index)")`: the string
        // hash's low bits track their input, so consecutive cards came out one
        // pixel apart -- a fixed-height grid wearing a masonry's clothes. The
        // spread is the whole point of the screen, because the slots keep
        // their heights and a card visibly resizes into the one it is dragged
        // over.
        heights = cards.map { index in
            80 + CGFloat(DemoTheme.stableMix(index) % 121)
        }

        let layout = UICollectionViewFlowLayout()
        layout.minimumInteritemSpacing = 10
        layout.minimumLineSpacing = 10
        layout.sectionInset = UIEdgeInsets(top: 12, left: 12, bottom: 12, right: 12)

        let collection = UICollectionView(frame: view.bounds, collectionViewLayout: layout)
        // The Objective-C never set this: UICollectionView used to default to a
        // black background, which is what separated the white cells in the
        // original demo. The cards now carry their own hairline and shadow, so
        // the background can be the ordinary grouped one.
        collection.backgroundColor = DemoTheme.Surface.background
        collection.delegate = self
        collection.dataSource = self
        collection.accessibilityIdentifier = "moodboard"
        collection.register(UICollectionViewCell.self, forCellWithReuseIdentifier: "CollectionViewCell")
        view.addSubview(collection)
        self.collection = collection

        collection.reloadData()
    }

    private func palette(for card: Int) -> Palette {
        SampleData.palettes[card % SampleData.palettes.count]
    }
}

// MARK: -

extension NormalCollectionViewController: UICollectionViewDataSource, UICollectionViewDelegate, UICollectionViewDelegateFlowLayout {

    func numberOfSections(in collectionView: UICollectionView) -> Int { 1 }

    func collectionView(_ collectionView: UICollectionView, numberOfItemsInSection section: Int) -> Int {
        cards.count
    }

    func collectionView(_ collectionView: UICollectionView,
                        cellForItemAt indexPath: IndexPath) -> UICollectionViewCell {

        let cell = collectionView.dequeueReusableCell(withReuseIdentifier: "CollectionViewCell",
                                                      for: indexPath)
        cell.backgroundColor = .clear
        cell.contentView.subviews.forEach { $0.removeFromSuperview() }

        let card = cards[indexPath.row]
        let palette = palette(for: card)
        let swatch = SwatchChip(color: UIColor(hex: palette.hex) ?? .systemGray,
                                caption: palette.name,
                                detail: palette.hex,
                                identifier: "swatch-\(card)")
        swatch.frame = cell.contentView.bounds
        swatch.autoresizingMask = [.flexibleWidth, .flexibleHeight]
        cell.contentView.addSubview(swatch)

        return cell
    }

    func collectionView(_ collectionView: UICollectionView,
                        willDisplay cell: UICollectionViewCell,
                        forItemAt indexPath: IndexPath) {
        // On the cell, not the swatch: the collection view resolves an index
        // path from the cell.
        collectionView.enableDragAndDrop(for: cell)
    }

    func collectionView(_ collectionView: UICollectionView,
                        layout collectionViewLayout: UICollectionViewLayout,
                        sizeForItemAt indexPath: IndexPath) -> CGSize {
        CGSize(width: (view.bounds.width / 2) - 17, height: heights[indexPath.row])
    }

    func collectionView(_ collectionView: UICollectionView,
                        moveItemAt sourceIndexPath: IndexPath,
                        to destinationIndexPath: IndexPath) {
        let card = cards.remove(at: sourceIndexPath.row)
        cards.insert(card, at: destinationIndexPath.row)
    }
}
